### PR TITLE
Added Reconnect support for Event Hubs

### PR DESCRIPTION
**Pull request by LarryOsterman**
Initial reconnect support

Test infrastructure in progress.

### CODE REVIEW

File: sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft Corporation. All Rights reserved
 // Licensed under the MIT license.
 
+// cspell: ignore Cloneable
```
Comment (heaths) [inline sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs]: Did this work? I thought there had to be no spaces e.g.,


```suggestion
// cspell:ignore Cloneable
```

File: Cargo.toml [hunk 2]
```diff
@@ -81,9 +81,9 @@ cargo_metadata = "0.18.1"
 clap = { version = "4.4.16", features = ["derive"] }
 criterion = { version = "0.5", features = ["async_tokio"] }
 dyn-clone = "1.0"
-fe2o3-amqp = { version = "0.14", features = ["native-tls", "tracing", "uuid"] }
+fe2o3-amqp = { version = "0.14", features = ["native-tls", "uuid"] }
 fe2o3-amqp-ext = { version = "0.14" }
-fe2o3-amqp-management = { version = "0.14", features = ["tracing"] }
+fe2o3-amqp-management = { version = "0.14", features = [] }
```
Comment (heaths) [inline Cargo.toml]: Nit:
```suggestion
fe2o3-amqp-management = { version = "0.14" }
```

File: sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs [hunk 3]
```diff
@@ -23,6 +24,8 @@ use azure_core_amqp::{
     AmqpManagement, AmqpReceiver, AmqpReceiverApis, AmqpReceiverOptions, AmqpSender,
     AmqpSenderApis, AmqpSession, AmqpSessionApis, AmqpSessionOptions, AmqpSource, AmqpSymbol,
 };
+#[cfg(feature = "test")]
```
Comment (heaths) [inline sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs]: Why `feature = "test"` and not just `test`?

File: sdk/eventhubs/azure_messaging_eventhubs/tests/force_errors.rs [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+                    let batch = producer
+                        .create_batch(Some(EventDataBatchOptions {
+                            partition_id: Some(EVENTHUB_PARTITION.to_string()),
+                            partition_key: Some("My Partition Key.".to_string()),
+                            ..Default::default()
+                        }))
+                        .await
+                        .unwrap();
+
+                    for i in 1..200 {
```
Comment (heaths) [inline sdk/eventhubs/azure_messaging_eventhubs/tests/force_errors.rs]: Given the 10, 30, even 45 second time outs, is this sample *always* going to be enough? Should this async fn actually sleep itself to make sure the producer is forced into an error in time? If I'm following the logic correctly, this seems to assume a lot about modern hardware (which, of course, will change).

**Review by heaths — CHANGES_REQUESTED** [review]

File: sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs (hunk 1 above)
Comment (LarryOsterman) [reply sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs]: It has in the past. And cspell doesn't seem to complain. I'll check.

File: sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs (hunk 3 above)
Comment (LarryOsterman) [reply sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs]: Because [test] does not work across package boundaries. I need a feature to enable test hooks that are used across package boundaries (or from tests in the `test` directory.

File: sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs (hunk 1 above)
Comment (LarryOsterman) [reply sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs]: And it appears that it's not actually used. So removing it.

File: sdk/eventhubs/azure_messaging_eventhubs/tests/force_errors.rs (hunk 4 above)
Comment (LarryOsterman) [reply sdk/eventhubs/azure_messaging_eventhubs/tests/force_errors.rs]: The test puts 200 messages in a batch and then loops continuously until the future is dropped.

The 200 is just the size of the batch being sent (lines 79-90 add events to the batch, and 91 sends the batch.

File: sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/sender.rs [hunk 5]
```diff
@@ -18,7 +19,7 @@ use tracing::{debug, warn};
 /// the retry functionality. That allows implementations which call into the Send API to not have
 /// to worry about retrying the operation themselves.
 pub(crate) struct RecoverableSender {
-    recoverable_connection: Arc<RecoverableConnection>,
+    recoverable_connection: Weak<RecoverableConnection>,
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/sender.rs:22]: This change is necessary to ensure that the sender does not keep a strong reference to the `RecoverableConnection` - that ensures that outstanding senders don't stop the ability to close a `RecoverableConnection`.

File: sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs [hunk 6]
```diff
@@ -59,14 +65,17 @@ pub(crate) struct RecoverableConnection {
...
+    connections: AsyncMutex<Option<Arc<AmqpConnection>>>,
+    mgmt_client: AsyncMutex<Option<Arc<AmqpManagement>>>,
     session_instances: AsyncMutex<HashMap<Url, Arc<AmqpSession>>>,
     receiver_instances: AsyncMutex<HashMap<Url, Arc<AmqpReceiver>>>,
     sender_instances: AsyncMutex<HashMap<Url, Arc<AmqpSender>>>,
     pub(super) authorizer: Arc<Authorizer>,
     connection_name: String,
     pub(super) retry_options: RetryOptions,
+
+    #[cfg(feature = "test")]
```
[the remark this answers is not in the document]
Comment (LarryOsterman) [reply sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs]: The `test` feature enables test hooks in the eventhubs codebase. Since the test which uses this code is outside the package (because it is an end-to-end test), it cannot use the `test` configuration.

File: sdk/core/azure_core_amqp/CHANGELOG.md [hunk 7]
```diff
@@ -1,5 +1,15 @@
 # Release History
 
+## 0.7.0 (Unreleased)
+
+### Features Added
+
+### Breaking Changes
+
+- Distinguish remote disconnect and remote closed errors by origin
```
Comment (heaths) [inline sdk/core/azure_core_amqp/CHANGELOG.md:9]: At face value this doesn't seem like a breaking change. If it is, cool. Just wanted to mention it in case it ended up under the wrong header.

Comment (LarryOsterman) [reply sdk/core/azure_core_amqp/CHANGELOG.md:9]: I added a bunch of fields to an enumeration which is not marked as non-exhaustive, which is a breaking change.

**Comment by LarryOsterman** [comment]
@heaths There was a merge conflict, can you please reapprove?
